src/rgw/rgw_rest_sts.cc

```diff
@@ -223,137 +224,265 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
   if (parser.parse(openidc_resp.c_str(), openidc_resp.length())) {
     JSONObj::data_val val;
     if (parser.get_data("jwks_uri", &val)) {
-      cert_url = val.str.c_str();
-      ldpp_dout(dpp, 20) << "Cert URL is: " << cert_url.c_str() << dendl;
+      jwks_url = val.str.c_str();
+      ldpp_dout(dpp, 20) << "JWKS URL is: " << jwks_url.c_str() << dendl;
     } else {
-      ldpp_dout(dpp, 0) << "Malformed json returned while fetching openidc url" << dendl;
+      ldpp_dout(dpp, 0) << "Malformed json returned while fetching jwks url" << dendl;
     }
   }
-  return cert_url;
+  return jwks_url;
+}
+
+vector<string>
+WebTokenEngine::get_x5c_certs_from_x5u_url(const DoutPrefixProvider* dpp, const string& x5u_url, optional_yield y) const
+{
+  bufferlist x5u_resp;
+  RGWHTTPTransceiver x5u_req(cct, "GET", x5u_url, &x5u_resp);
+  //Headers
+  x5u_req.append_header("Content-Type", "application/x-www-form-urlencoded");
+
+  int res = x5u_req.process(y);
+  if (res < 0) {
+    ldpp_dout(dpp, 10) << "HTTP request res: " << res << " for getting x5c from x5u" << dendl;
+    throw -EINVAL;
+  }
+  string certs = x5u_resp.c_str();
+  vector<string> x5c;
+  const regex pattern("(-----BEGIN CERTIFICATE-----(.*?)-----END CERTIFICATE-----)+");
+  for(sregex_iterator it = sregex_iterator(
+      certs.begin(), certs.end(), pattern);
+      it != sregex_iterator(); it++) {
+      smatch match;
+      match = *it;
+      x5c.emplace_back(match.str(1));
+      ldpp_dout(dpp, 10) << "Matched: " << match.str(1) << dendl;
+  }
+  return x5c;
+}
+
+void
+WebTokenEngine::validate_signature_using_cert(const DoutPrefixProvider* dpp, const jwt::decoded_jwt& decoded, const string& algorithm, const vector<string>& certs, const vector<string>& thumbprints, bool add_pem_str) const
+{
+  string cert;
+  bool found_valid_cert = false;
+  for (auto& it : certs) {
+    if (add_pem_str) {
+      cert = "-----BEGIN CERTIFICATE-----\n" + it + "\n-----END CERTIFICATE-----";
+    } else {
+      cert = it;
+    }
+    if (is_cert_valid(thumbprints, cert)) {
+      found_valid_cert = true;
```

Comment:
found_valid_cert = get true value if we enter the previous loop?  and in line 278 again it will be true?

Reply:
Can you elaborate on that? I don't think I see the issue. It looks like it's false when we enter the loop and breaks out as soon as it's set true, where it will be at 278.

Reply:
In the version he was looking at, there was an extra line "found_valid_cert = true;" at the end of the loop.  That means if there was more than one cert (likely always the case) it would always think it had found a trust anchor.  The latest version that pritha pushed removed that line.

If I understand the code that's there now, it will correctly validate that one of the certificates is a trust anchor, but it doesn't validate the chain to see if each signed the previous certificate.    If that's what it's doing, then this logic could be fooled by supplying cert[0] = bad-guy, cert[1]=trusted, cert[0] not signed by cert[1].